### PR TITLE
[CARBONDATA-2809][DataMap] Block rebuilding for bloom/lucene and preagg datamap

### PR DESCRIPTION
**Pull request by xuchuanyin**
# Problems & RootCause:

For non-lazy (not deferred rebuild, which is by default) index datamap, the data of datamap will be
generated immediately after:
1. the datamap is created
2. the main table is loaded
So there is no need to rebuild this datamap.
Actually, it will encounter error if we trigger rebuilding for these
datamaps due to the existence of old data.

The situation for preagg is the same.

# Solution:

We will block rebuilding for bloom/lucene and preagg datamap as well as creating them with 'deferred rebuild' in carbondata 1.4.1.
Later we will optimize the full rebuilding and incremental rebuilding for these datamaps.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 `NO`
 - [x] Any backward compatibility impacted?
  `NO`
 - [x] Document update required?
 `NO`
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
 `NO`
        - How it is tested? Please attach test report.
`Tested in local`
        - Is it a performance related change? Please attach the performance test report.
 `NO`
        - Any additional information to help reviewers in testing this change.
        `NA`
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
`NA`

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7706/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6432/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6098/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6108/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7715/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6441/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRebuildCommand.scala [hunk 1]
```diff
@@ -48,7 +50,17 @@ case class CarbonDataMapRebuildCommand(
         )(sparkSession)
     }
     val provider = DataMapManager.get().getDataMapProvider(table, schema, sparkSession)
-    provider.rebuild()
+    // for non-lazy index datamap, the data of datamap will be generated immediately after
+    // the datamap is created or the main table is loaded, so there is no need to
+    // rebuild this datamap.
+    if (!schema.isLazy && provider.isInstanceOf[IndexDataMapProvider]) {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRebuildCommand.scala]: Even if it is other datamap like preaggregat datamap, we should not rebuild it, right?

Comment (xuchuanyin) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRebuildCommand.scala]: For MV, current implementation requires rebuild.
For preagg, I'm not sure about its implementation, so leave it as it is.

Comment (xuchuanyin) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonDataMapRebuildCommand.scala]: OK.

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6489/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7765/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6147/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6148/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6177/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6178/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7799/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6523/

**Comment by ravipesala** [comment]
@xuchuanyin Please check MVTests, it is failing

**Comment by xuchuanyin** [comment]
@ravipesala 
Fixed.
The root cause is that MV is actually 'deferred rebuild', but we didn't specify it while we create the datamap.
To make compliance, we will enable 'deferred rebuild' for MV datamap no matter whether the flag is enabled by user or not.

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6182/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7803/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6527/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7806/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6530/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**